tests/unit/channel_test.py

```diff
@@ -68,3 +72,82 @@ def test_read_bulk_response(ssh_client_session):
             break
 
     assert b'Hello World' in response  # noqa: WPS302
+
+
+def test_request_exec(ssh_channel):
+    """Test direct call to request_exec."""
+    ssh_channel.request_exec('exit 1')
+
+    rc = -1
+    while rc == -1:
+        ssh_channel.poll(timeout=5)
+        rc = ssh_channel.get_channel_exit_status()
+    assert rc == 1
+
+
+def test_send_eof(ssh_channel):
+    """Test send_eof correctly terminates input stream."""
+    ssh_channel.request_exec('cat')
+    ssh_channel.send_eof()
+
+    rc = -1
+    while rc == -1:
+        ssh_channel.poll(timeout=5)
+        rc = ssh_channel.get_channel_exit_status()
+    assert rc == 0
+
+
+def test_send_signal(ssh_channel):
+    """Test send_signal correctly forwards signal to the process."""
+    ssh_channel.request_exec('bash -c \'trap "exit 1" SIGUSR1; echo ready; sleep 5; exit 0\'')
+
+    # Wait until the process is ready to receive signal
+    output = ''
+    while not output.startswith('ready'):
+        ssh_channel.poll(timeout=1000)
+        output += ssh_channel.recv().decode('utf-8')
+
+    # Send SIGINT
+    ssh_channel.send_signal('USR1')
+
+    rc = -1
+    while rc == -1:
+        ssh_channel.poll(timeout=5)
+        rc = ssh_channel.get_channel_exit_status()
+
+    assert rc == 1
+
+
+def test_recv_eof(ssh_channel):
+    """
+    Test that reading EOF does not raise error.
+
+    SystemError: Negative size passed to PyBytes_FromStringAndSize
+    """
+    ssh_channel.request_exec('exit 0')
+    ssh_channel.poll(timeout=1000)
+    ssh_channel.recv()
+
+
+def test_is_eof(ssh_channel):
+    """Test that EOF-state is correctly obtained with is_eof."""
+    ssh_channel.request_exec('exit 0')
+    ssh_channel.poll(timeout=5000)  # noqa: WPS432
+    assert ssh_channel.is_eof()
+
+
+def test_destructor(sshd_addr, ssh_clientkey_path):
+    """
+    Garbage collector can destroy session before channel.
+
+    Test that this event does not cause a segfault in channels destructor.
+    """
+    def _do_not_crash():  # noqa: WPS430
+        ssh_session = Session()
+        ensure_ssh_session_connected(ssh_session, sshd_addr, ssh_clientkey_path)
+        ssh_channel = ssh_session.new_channel()  # noqa: F841
+
+    # Without fix, garbage collector first deletes session and we segfault
+    # in channel destructor when trying to access low-level C session object.
+    _do_not_crash()
+    gc.collect(0)
```

Comment:
It should be more stable if you first disable GC, then collect everything, and only then run the test helper, finally performing the collection call that checks for crashes. You could also assert the number of objects, maybe. Before and after. Then, re-enable GC upon completion of the test.

You could integrate `pypytools` and make a helper like https://github.com/cherrypy/cheroot/blob/0fd16f0c3f111c3709a4d14e6400cabf2c094239/cheroot/test/test_server.py#L412-L416 but that's not mandatory. Manual invocations in just one place is fine.

Reply:
I can't assert number of objects since it can be easily changed. Even `gc.is_tracked` can not be asserted as for some reason, it returns False for ssh_channel, but True for ssh_session. But I added `gc.disable` to have a better control over it.

Anyway, the crash is always triggered by the test code. Feel free to suggest improvements if you see any.